### DISCUSSION TITLE
React Native Support

### DISCUSSION THREAD

**Discussion by MaximoLiberata** [mqttjs/MQTT.js#1841] [General]
In the version [5.5.2](https://github.com/mqttjs/MQTT.js/releases/tag/v5.5.2) was fixed an issue that didn't allow use MQTT.js in React Native (RN). The motivation of this discussion is bring a example of how to implement MQTT.js in React Native and also maybe bring ideas of enhancements for RN environment.

You can see the example in [React Native MQTT.js Implementation](https://github.com/MaximoLiberata/react-native-mqtt.js-example)

**Comment by robertsLando** [comment]
@MaximoLiberata Thanks! Could you kindly add a link to this on readme please?

**Reply by MaximoLiberata** [reply]
Hi @robertsLando, thanks to you!

I opened the PR [#1842 ](https://github.com/mqttjs/MQTT.js/pull/1842)

**Comment by robertsLando** [comment]
Wondering if this is still needed? https://github.com/MaximoLiberata/react-native-mqtt.js-example/blob/main/App.js#L8

**Reply by MaximoLiberata** [reply]
Yes, is still needed, I'm searching the reason

**Reply by robertsLando** [reply]
Strange that the fix isn't picked up by  https://github.com/mqttjs/MQTT.js/blob/78e4f873a0575bc700a2c7d0ec6728cd86671ebc/src/lib/shared.ts#L79-L84

**Reply by MaximoLiberata** [reply]
Is not fixed there, I edited that because RN has a process but not a process.nextTick and also a prevention in other platform implementation, but that doesn't fix the global process.nextTick

**Reply by MaximoLiberata** [reply]
As said @abdelhameedhamdy in [#1840](https://github.com/mqttjs/MQTT.js/pull/1840#issuecomment-2052501909) I found the same error with the package `process-nextick-args`, it's a dependencie of other packages as: `mqtt-packet` and `readable-stream`, process.nextTick is used in global scope for any package

**Reply by MaximoLiberata** [reply]
If we modify the package `process-nextick-args` and instead of using `process.nextTick` we replace it for the following function `var fnNextTick = typeof process?.nextTick === 'function' ? process.nextTick : (callback) => {
 setTimeout(callback, 0)
 }`, the error is fixed and isn't necessary to add `process.nextTick` at entry file of RN.

I don't know if this is the best solution, @robertsLando What do you think?

**Reply by abdelhameedhamdy** [reply]
it seems that polyfill with [setImmediate](https://facebook.github.io/react-native/docs/timers.html) fix this bug as well,

By Adding this at the top of app 

```
process.nextTick = setImmediate 
```
like 
```
process.nextTick = (callback) => {
  setTimeout(callback, 0);
}
```

**Reply by robertsLando** [reply]
Just to be sure, is react native using the browser bundle? If so I think this could be easily fixed by patching the esbuild bundle here: https://github.com/mqttjs/MQTT.js/blob/main/esbuild.js

Maybe by adding a new entry on "define" or manually adding the polifilly on top of the output bundfle, could you guys do some tries?

**Reply by MaximoLiberata** [reply]
I tried to use the browser bundle, but they threw a exeption:

`import * as mqtt from 'mqtt/dist/mqtt'` &rarr; Return {"default": {}}
`import * as mqtt from 'mqtt/dist/mqtt.min'` &rarr; Return {"default": {}}
`import mqtt from 'mqtt/dist/mqtt.esm'` &rarr; Throw Error: ws does not work in the browser. Browser clients must use the native WebSocket object

React Native doesn't use the browser bundle, I think its behavior is closer to node.js than browser.

Also, I tried to add a piece of code in the `src/index.ts` file and worked:

```
import { isReactNativeBrowser } from './lib/is-browser'

import * as mqtt from './mqtt'

if (isReactNativeBrowser && typeof process?.nextTick !== 'function') {
	process.nextTick = (callback: () => void) => {
		setTimeout(callback, 0)
	}
}

export default mqtt
export * from './mqtt'

```

Not is the best solution, this change also modify the `process` object in the global scope.

**Reply by abdelhameedhamdy** [reply]
@MaximoLiberata 

I just tried adding this block

```
if (typeof process?.nextTick !== 'function') {
	process.nextTick = setImmediate;
}
```
or, same like

```
if (typeof process?.nextTick !== 'function') {
	process.nextTick = (callback: () => void) => {
		setTimeout(callback, 0)
	}
}
```

in https://github.com/mqttjs/MQTT.js/blob/main/src/lib/connect/index.ts#L11

I think it fixed it, without needs to add it at top of RN App.

If you can have a look to check if that might be a proper solution.

**Reply by abdelhameedhamdy** [reply]
@MaximoLiberata Did the last proposed solution fix the issue!

**Reply by MaximoLiberata** [reply]
Yes, I tested in the RN APP example and it worked well.

**Comment by abdelhameedhamdy** [comment]
Good, I think this update might work with release [v5.5.1](https://github.com/mqttjs/MQTT.js/releases/tag/v5.5.1) as a fix, I am not sure, needs to best tested.
If it did not, so v5.5.2 needs to add this update with a new PR in your current RN mqtt.js fork.

**Comment by robertsLando** [comment]
> I tried to use the browser bundle, but they threw a exeption:

Nope what I meant is if react is actually using the browser bundle or not, the bundle is automatically served based on the package.json `exports` field if it detects a browser env, otherwise the "node" version is serverd.

I think that adding:

```
if (isReactNativeBrowser && typeof process?.nextTick !== 'function') {
	process.nextTick = (callback: () => void) => {
		setTimeout(callback, 0)
	}
}
```

to the connect/index.ts isn't so bad, someone wants to submit a PR?

**Reply by abdelhameedhamdy** [reply]
@robertsLando 
I just tested it vs [v5.5.1](https://github.com/mqttjs/MQTT.js/releases/tag/v5.5.1), it works well with just updating connect/index.ts, and only the first commit by @MaximoLiberata in [BufferedDuplex.ts](https://github.com/mqttjs/MQTT.js/blob/fc8fafbdf5e01edc487192393293b944e77f5920/src/lib/BufferedDuplex.ts#L2) 
which needs importing Buffer package.

So, I am not sure that other commits in ([fc8fafb](https://github.com/mqttjs/MQTT.js/commit/fc8fafbdf5e01edc487192393293b944e77f5920)) are required, as no error appears for me at least. may be required for other issue.

just this block in connect/index.ts, and importing Buffer
```
if (typeof process?.nextTick !== 'function') {
	process.nextTick = setImmediate;
}
```
even no need for importing isReactNativeBrowser() in connect/index.ts

with regards to using setImmediate or setTimeout(callback, 0)
I searched for the difference between both, and it seems that it is [preferred](https://medium.com/@webtaculars/settimeout-vs-setimmediate-6502fae56a70) to use setImmediate.

**Reply by abdelhameedhamdy** [reply]
Just to update, my issue was that process.nextTick is not a function was blocking my RN App to connect to the broker.
([fc8fafb](https://github.com/mqttjs/MQTT.js/commit/fc8fafbdf5e01edc487192393293b944e77f5920)) was solving a different issue other than mine, so the new proposed fix for resolving the issue to add a block of code at start of RN App to let it work.

So, just a PR is required to fix the process.nextTick is not a function error for RN Apps.

**Reply by abdelhameedhamdy** [reply]
Another notice here, I did not get this error: Cannot create URL for blob (https://github.com/mqttjs/MQTT.js/pull/1840) in my RN App; as I am setting timerVariant: "native" not "auto".

**Reply by robertsLando** [reply]
> Another notice here, I did not get this error: Cannot create URL for blob (https://github.com/mqttjs/MQTT.js/pull/1840) in my RN App; as I am setting timerVariant: "native" not "auto".

Yeah that's the solution I suggested at start in the issue

**Reply by abdelhameedhamdy** [reply]
@robertsLando I just opened PR [#1849](https://github.com/mqttjs/MQTT.js/pull/1849).
